новые методы для работы со временем в InMemoryTaskManager;

AlexMarchz opened the pull request:
расширение форматирования для работы со временем в TaskUtils; Обновление тестов;
Добавление новых полей в data классах для времени.

AlexMarchz committed:
новые методы для работы со временем в InMemoryTaskManager;
расширение форматирования для работы со временем в TaskUtils;
Обновление тестов;
Добавление новых полей в data классах для времени.

AlexMarchz committed:
Codestyle

Ivanov-s-tmn commented on src/data/Task.java:
```diff
@@ -23,11 +29,42 @@ public Task(String name, String descriptions, Status status) {
+        this.duration = duration;
+    }
+
+    public Task(String description, String name) {
+        this.name = name;
+        this.descriptions = description;
+        status = Status.NEW;
+    }
```
Из одного конструктора можно вызвать другой, чтобы сократить дублирование инициализации полей, как пример: `this(name, description, TaskStatus.NEW);`

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
Не забывай, что из файла теперь еще нужно задачи и подзадачи закинуть в отсортированный список

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -26,10 +31,92 @@ private int generatorId() {
         return nextId++;
     }
 
+    public void updateEpicTime(Epic epic) {
+        List<Task> subTasksList = getSubTasksForEpic(epic);
+        if (subTasksList.isEmpty()) {
+            return;
+        }
+
+        Duration totalDuration = Duration.ofMinutes(0);
+        LocalDateTime startTime = subTasksList.getFirst().getStartTime(); // Время начала
+        LocalDateTime endTime = subTasksList.getLast().getEndTime(); // Время окончания
+
+        for (Task subTask : subTasksList) {
+            totalDuration = totalDuration.plus(subTask.getDuration());
+            if (subTask.getStartTime().isBefore(startTime)) {
+                startTime = subTask.getStartTime();  // Обновляем время
+            }
+            if (subTask.getEndTime().isAfter(endTime)) {
+                endTime = subTask.getEndTime();  // Обновляем время
+            }
+        }
+
+        epic.setStartTime(startTime);
+        epic.setEndTime(endTime);
+        epic.setDuration(totalDuration);
+    }
```
Сделай метод приватным, он используется как и метод обновления статуса только в логике других методов этого класса

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -26,10 +31,92 @@ private int generatorId() {
+        LocalDateTime startTime = subTasksList.getFirst().getStartTime(); // Время начала
+        LocalDateTime endTime = subTasksList.getLast().getEndTime(); // Время окончания
+
+        for (Task subTask : subTasksList) {
+            totalDuration = totalDuration.plus(subTask.getDuration());
+            if (subTask.getStartTime().isBefore(startTime)) {
+                startTime = subTask.getStartTime();  // Обновляем время
+            }
+            if (subTask.getEndTime().isAfter(endTime)) {
+                endTime = subTask.getEndTime();  // Обновляем время
+            }
+        }
```
Любое временное поле сабтаски может быть null, этот момент нужно обработать, чтобы не упасть с NPE

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -26,10 +31,92 @@ private int generatorId() {
+        return subTasks;
+    }
+
+    public void addPriority(Task task) {
+        if (task.getType().equals(TaskType.EPIC)) {
+            return;  // Не добавляем эпики
+        }
+
+        if (task.getStartTime() != null && task.getEndTime() != null) {
+            for (Task existingTask : priority) {
+                if (existingTask.getId() == task.getId()) {
+                    priority.remove(existingTask);  // Удаляем старую версию, если она уже есть
+                }
+                if (checkForIntersection(task, existingTask)) {
+                    return;  // Если есть пересечение, не добавляем задачу
+                }
+            }
+            priority.add(task);  // Если пересечений нет, добавляем задачу
+        }
+    }
```
Метод приватный

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -26,10 +31,92 @@ private int generatorId() {
+                !task1.getStartTime().isAfter(task2.getEndTime());
+    }
+
+    public void relevancePriority(Task task) {
+        if (task == null || task.getStartTime() == null) {
+            return;
+        }
+
+        for (Task existingTask : priority) {
+            if (existingTask == task) {
+                continue;
+            }
+            if (checkForIntersection(task, existingTask)) {
+                throw new ManagerSaveException("Задачи " + task.getId() + " и " + existingTask.getId() + " пересекаются");
+            }
+        }
+    }
```
Этот тоже приватный

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -26,10 +31,92 @@ private int generatorId() {
+        return subTasks;
+    }
+
+    public void addPriority(Task task) {
+        if (task.getType().equals(TaskType.EPIC)) {
+            return;  // Не добавляем эпики
+        }
+
+        if (task.getStartTime() != null && task.getEndTime() != null) {
+            for (Task existingTask : priority) {
+                if (existingTask.getId() == task.getId()) {
+                    priority.remove(existingTask);  // Удаляем старую версию, если она уже есть
+                }
+                if (checkForIntersection(task, existingTask)) {
+                    return;  // Если есть пересечение, не добавляем задачу
+                }
+            }
+            priority.add(task);  // Если пересечений нет, добавляем задачу
+        }
```
Зачем дублировать проверки на пересечение, если это уже есть в методе `relevancePriority()`

Ivanov-s-tmn commented on test/manager/InMemoryTaskManagerTest.java:
```diff
@@ -3,12 +3,15 @@
+import java.time.LocalDateTime;
 import java.util.List;
 
 class InMemoryTaskManagerTest {
```
Смотри. для тестов таск менеджера и файлового менеджера нужно реализовать абстрактный базовый класс, в котором будет реализация всех тестов для методов интерфейса `TaskManager`

Благодаря параметризации `<T extends TaskManager>` мы задаем для наследников `TaskManagerTest` ограничения по типам, которые может принимать параметризованное поле (т.е. это наследники `TaskManager`)
```java
public abstract class TaskManagerTest<T extends TaskManager> {
     protected T taskManager; //параметризованное поле taskManager, которое мы можем использовать в наследниках и инициализировать его конкретной реализацией таск менеджера

     /**
     * Реализация тестов всех методов интерфейса `TaskManager` должна быть в этом классе, классы-наследники уже будут их выполнять с конкретной реализацией менеджера
     */

}
```
Как пример:
```java
public class InMemoryTaskManagerTest extends TaskManagerTest<InMemoryTaskManager> {

    @BeforeEach
    void beforeEach() {
        taskManager = new InMemoryTaskManager(); //инициализируем T taskManager реализацией InMemoryTaskManager
        initTasks(); //Можно вынести создание тестовых задач в TaskManagerTest
    }
}
```

С тестами файлового менеджера ситуация аналогичная, только инициализировать таск менеджер будет реализацией `FileBackedTaskManager` и будут дополнительные тесты его функционала (проверка восстановления состояния из файла). `InMemoryTaskManagerTest` по идее может быть пустым и прогонять тесты `TaskManagerTest`, потому что в нем нет особого функционала как у файлового менеджера

В этом классе нужно покрыть тестами все методы интерфейса `TaskManager`, они будут автоматически прогонятся в реализациях (ин мемори менеджере и файловом менеджере), в них только будут реализованы тесты для функционала, который является уникальным для конкретной реализации менеджера

AlexMarchz committed:
Абстрактный класс TaskManagerTest,
запись в priority из файла;

Ivanov-s-tmn commented on test/manager/TaskManagerTest.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    void testShowAllSubTasks() {                                           //Сергей, привет, поставил 0 потому что не смог разобраться с тем, как организовать сетап, если передавать
+        taskManager.addEpic(epic);                                        //сабтаску через epic.getEpicId то тут всё ок, но loadFromFile 0 сабов будет иметь по факту, а если
+        taskManager.addSubTask(subTask);                                 //сабтаску передавать единицу, то тут по факту 0, зато там всё как и должно было быть
+        assertEquals(0, taskManager.showAllSubTasks().size()); //после единого сетапа не нашёл вариантов, пытался разные initTasks сделать и проверки, но бестоклу
```
У тебя по идее счетчик айди с нуля идет, поэтому эпику присваивается айди 0 после вызова `addEpic()`, по факту у эпика должен быть айди 1, а у сабтаски 2. У тебя так же могут и не работать тесты где еще добавляются эпик и сабтаск (например в удалении и сабтасок), но сейчас они проходят потому что нет проверки, что подзадача создалась и добавилась

AlexMarchz committed:
Правка по тестам: отдельные проверки добавления и поиска по id;
showAllSubTask коректно проверяет с единым сетапом для двух классов;
id теперь инкриминируется с единицы;

AlexMarchz committed:
CodeStyle бесполезные импорты
